avocado/core/mux.py

```diff
@@ -29,6 +36,152 @@
 REMOVE_VALUE = 1
 
 
+class MuxTree(object):
+
+    """
+    Object representing part of the tree from the root to leaves or another
+    multiplex domain. Recursively it creates multiplexed variants of the full
+    tree.
+    """
+
+    def __init__(self, root):
+        """
+        :param root: Root of this tree slice
+        """
+        self.pools = []
+        for node in self._iter_mux_leaves(root):
+            if node.is_leaf:
+                self.pools.append(node)
+            else:
+                self.pools.append([MuxTree(child) for child in node.children])
+
+    @staticmethod
+    def _iter_mux_leaves(node):
+        """ yield leaves or muxes of the tree """
+        queue = collections.deque()
+        while node is not None:
+            if node.is_leaf or getattr(node, "multiplex", None):
+                yield node
+            else:
+                queue.extendleft(reversed(node.children))
+            try:
+                node = queue.popleft()
+            except IndexError:
+                raise StopIteration
+
+    def __iter__(self):
+        """
+        Iterates through variants
+        """
+        pools = []
+        for pool in self.pools:
+            if isinstance(pool, list):
+                pools.append(itertools.chain(*pool))
+            else:
+                pools.append(pool)
+        pools = itertools.product(*pools)
+        while True:
+            # TODO: Implement 2nd level filters here
+            # TODO: This part takes most of the time, optimize it
+            yield list(itertools.chain(*pools.next()))
+
+
+class MuxPlugin(object):
+    """
+    Follows the Multiplexer API to produce variants
```

Comment:
Multiplexer API is something abstract at this point and definitely nothing people should be using. Anyway one of the requirements was that the multiplexer idea should be part of the core so people could inspire and based their plugins on it. That's why I created this more-less generic class which is then used by the actual yaml_to_mux plugin.

Anyway to your question, the current API is the `__init__` where you hand over the `root` and other useful values and this class will implement the rest of the necessary things in order to produce complete varianter plugin.